be/src/exec/pipeline/schedule/observer.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "exec/pipeline/schedule/observer.h"
+
+#include "exec/pipeline/pipeline_driver.h"
+#include "exec/pipeline/schedule/common.h"
+
+namespace starrocks::pipeline {
+static void on_update(PipelineDriver* driver) {
+    auto sink = driver->sink_operator();
+    auto source = driver->source_operator();
+    TRACE_SCHEDULE_LOG << "notify driver:" << driver << " state:" << driver->driver_state()
+                       << " in_block_queue:" << driver->is_in_block_queue()
+                       << " source finished:" << source->is_finished()
+                       << " operator has output:" << source->has_output() << " sink finished:" << sink->is_finished()
+                       << " sink need input:" << sink->need_input() << ":" << driver->to_readable_string();
+    if (sink->is_finished() || sink->need_input() || source->is_finished() || source->has_output()) {
+        driver->fragment_ctx()->event_scheduler()->try_schedule(driver);
+    }
+}
+
+static void on_sink_update(PipelineDriver* driver) {
+    auto sink = driver->sink_operator();
+    if (sink->is_finished() || sink->need_input()) {
+        driver->fragment_ctx()->event_scheduler()->try_schedule(driver);
+    }
+}
+
+static void on_source_update(PipelineDriver* driver) {
+    auto source = driver->source_operator();
+    if (source->is_finished() || source->has_output()) {
+        driver->fragment_ctx()->event_scheduler()->try_schedule(driver);
+    }
+}
+
+void PipelineObserver::_do_update(int event) {
+    auto driver = _driver;
+    auto token = driver->acquire_schedule_token();
+
+    if (driver->is_in_block_queue()) {
+        bool pipeline_block = driver->driver_state() != DriverState::INPUT_EMPTY ||
+                              driver->driver_state() != DriverState::OUTPUT_FULL;
```

Comment:
DriverState::PRECONDITION is neglected?

Reply:
This code is for special handling of PRECONDITION. Because in PRECONDITION state, has_output need_input may return false. In this case, we need to schedule the driver to INPUT_EMPTY/OUTPUT_FULL state.

I'll comment here.